### PR TITLE
Use static libs for CI ndebug build

### PR DESCRIPTION
**Pull request by sethrj**
See https://github.com/celeritas-project/celeritas/pull/233#issuecomment-837383052 for context

- Format class data/method declarations in header files for interactors and field
- Add missing `inline` keywords
- Use `STATIC` libraries for `cuda-ndebug` build on CI

### CODE REVIEW

**Comment by sethrj** [comment]
Ugh @pcanal you were right about static build issues ... if we'd tried this earlier, it might've given us some better heads-up for all the mysterious linking order issues and load-time crashes from CUDA RDC. 

The [CI build log](https://cloud.cees.ornl.gov/jenkins-ci/blue/rest/organizations/jenkins/pipelines/Celeritas/branches/PR-238/runs/2/nodes/11/steps/49/log/?start=0) is filled with link errors:
```
: && /usr/bin/c++ -fPIC -Wall -Wextra -pedantic -Werror  -std=c++14 -pipe -fsigned-char -pthread -O3 -DNDEBUG  -Wl,--no-as-needed   -pthread -shared -Wl,-soname,libceleritas_root.so -o src/libceleritas_root.so src/CMakeFiles/celeritas_root.dir/CeleritasRootInterface.cxx.o src/CMakeFiles/celeritas_root.dir/io/RootImporter.cc.o CMakeFiles/celeritas_root.dir/cmake_device_link.o  -Wl,-rpath,/opt/vecgeom/lib:/opt/software/linux-ubuntu20.04-x86_64/gcc-9.3.0/xerces-c-3.2.3-2fkvy47yjazu7dytx4iuthdxrsa2kvcx/lib  src/libceleritas.a  /opt/view/lib/libTree.so.6.20.08  /opt/view/lib/libHepMC3.so  /opt/view/lib/libmpi.so  /opt/vecgeom/lib/libvecgeomcuda.so  /opt/vecgeom/lib/libvgdml.so  /opt/software/linux-ubuntu20.04-x86_64/gcc-9.3.0/xerces-c-3.2.3-2fkvy47yjazu7dytx4iuthdxrsa2kvcx/lib/libxerces-c.so  /opt/vecgeom/lib/libvecgeom.so  -ldl  -lrt  -lpthread  /usr/local/cuda/lib64/libcudart.so  /opt/vecgeom/lib/libvecgeomcuda_static.a  /opt/view/lib/libNet.so.6.20.08  /opt/view/lib/libRIO.so.6.20.08  /opt/view/lib/libThread.so.6.20.08  /opt/view/lib/libCore.so.6.20.08  -lpthread  -lcudadevrt  -lcudart_static  -lrt  -lpthread  -ldl && :
/usr/bin/ld: src/libceleritas.a(Logger.cc.o): relocation R_X86_64_PC32 against symbol `_ZSt4cerr@@GLIBCXX_3.4' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
```

File: src/physics/em/detail/RayleighInteractor.i.hh [hunk 1]
```diff
@@ -31,8 +31,6 @@ RayleighInteractor::RayleighInteractor(const RayleighNativeRef& shared,
     , inc_direction_(direction)
     , element_id_(el_id)
 {
-    CELER_EXPECT(inc_energy_ >= this->min_incident_energy()
```
Comment (pcanal) [inline src/physics/em/detail/RayleighInteractor.i.hh:34]: Is this removal intentional?  It does not seem to be noted in the git log.

Comment (sethrj) [reply src/physics/em/detail/RayleighInteractor.i.hh:34]: Ah I removed that because inferred the min/max energy accessors from Rayleigh, which originally were in the other interactors as well

**Comment by sethrj** [comment]
@pcanal Can you please take a look at this branch? Even after rebasing on the fixed RDC stuff, I'm still seeing link errors. I think `NOT BUILD_SHARED_LIBS` in `celeritas_add_library` needs to go, and some of the libraries adapted based on whether the particular library is shared.
```
[167/275] Building CUDA object src/CMakeFiles/celeritas.dir/sim/detail/InitializeTracks.cu.o
FAILED: src/CMakeFiles/celeritas.dir/sim/detail/InitializeTracks.cu.o
/usr/local/cuda-10.2/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -DJSON_USE_IMPLICIT_CONVERSIONS=1 -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -DVECCORE_ENABLE_CUDA -D_MPICC_H -I/projects/spack/opt/spack/gcc-8.3.1/clhep/t4e47pz/lib/CLHEP-2.4.1.3/../../include -I../src -Isrc -isystem=/projects/spack/var/spack/environments/celeritas/.spack-env/view/include -isystem=/usr/local/cuda-10.2/include -isystem=/projects/spack/opt/spack/gcc-8.3.1/xerces-c/3goe2vs/include -isystem=/projects/spack/opt/spack/gcc-8.3.1/openmpi/n54mlfg/include -g -G --generate-code=arch=compute_35,code=[compute_35,sm_35] -Xcompiler -pthread -std=c++14 -MD -MT src/CMakeFiles/celeritas.dir/sim/detail/InitializeTracks.cu.o -MF src/CMakeFiles/celeritas.dir/sim/detail/InitializeTracks.cu.o.d -x cu -c ../src/sim/detail/InitializeTracks.cu -o src/CMakeFiles/celeritas.dir/sim/detail/InitializeTracks.cu.o
ptxas fatal   : Unresolved extern function '_ZN7vecgeom20globaldevicegeomdata11GetNavIndexEv'
```

**Comment by pcanal** [comment]
I'll check.

**Comment by pcanal** [comment]
Well ... it tooks more than 5 minutes ... https://github.com/celeritas-project/celeritas/pull/242

**Comment by sethrj** [comment]
Cherry-picking #242, the branch fails on emmet but passes on the CI, so static builds aren't entirely stable...

```
$ ninja
[2/95] Linking CXX executable CMakeFiles/base_Range.dir/cmake_device_link.o
FAILED: CMakeFiles/base_Range.dir/cmake_device_link.o
/usr/local/cuda-10.2/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -g -G --generate-code=arch=compute_35,code=[compute_35,sm_35] /home/s3j/.local/src/celeritas/build-static/src/libceleritas.a /projects/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvecgeomcuda_static.a -Xcompiler=-pthread -Xcompiler=-fPIC -Wno-deprecated-gpu-targets -shared -dlink test/CMakeFiles/base_Range.dir/base/Range.test.cc.o test/CMakeFiles/base_Range.dir/base/Range.test.cu.o -o CMakeFiles/base_Range.dir/cmake_device_link.o -L/usr/local/cuda-10.2/targets/x86_64-linux/lib/stubs  -L/usr/local/cuda-10.2/targets/x86_64-linux/lib test/libCeleritasTest.a src/libceleritas.a  -ldl -lrt -lpthread -lcudadevrt -lcudart_static
nvlink error   : Multiple definition of '_ZN9celeritas6detail71_GLOBAL__N__47_tmpxft_000fde2d_00000000_6_BetheHeitler_cpp1_ii_e3f3b31229bethe_heitler_interact_kernelENS0_20BetheHeitlerPointersENS_17ModelInteractRefsILNS_8MemSpaceE1EEE' in 'src/libceleritas.a:BetheHeitler.cu.o', first defined in '/home/s3j/.local/src/celeritas/build-static/src/libceleritas.a:BetheHeitler.cu.o'
nvlink fatal   : merge_elf failed
```

I'm assuming this is more CUDA compiler issues since we don't see it on the CUDA11 CI. I'll add some CMake code to fail fast if building static + cuda < 11 + vecgeom.

**Comment by pcanal** [comment]
I am not sure.  I notice that there is 2 *distinct* listing of the static library:
```
 /home/s3j/.local/src/celeritas/build-static/src/libceleritas.a 
...
src/libceleritas.a 
```
I am expecting multiple listing but with the same "name".

**Comment by sethrj** [comment]
Not sure if this matters, but setting `_midsuf` to a non-empty value (trying to debug by doing that) breaks the build due to the inability to install/alias the non-suffixed library 😒 Anyway, I think we can leave the full static library debugging for another day, perhaps after we have a conversation with Kitware and Nvidia. I think this is ready for merge as-is since it at least will catch some of the "missing inline" type errors on the current master.

**Comment by sethrj** [comment]
@pcanal I'm past ready to move on with life and forget about emmet compatibility, but here's the build sequence with static libs after changing the suffix to `_mid`, building (until failure), then deleting all the `.a` files:
```console
$ ninja -j1 -v demo-rasterizer
[1/5] : && /projects/spack/opt/spack/gcc-8.3.1/cmake/547luje/bin/cmake -E rm -f src/libceleritas_mid.a && /usr/bin/ar qc src/libceleritas_mid.a  src/CMakeFiles/celeritas_objects.dir/celeritas_version.cc.o src/CMakeFiles/celeritas_objects.dir/base/Assert.cc.o src/CMakeFiles/celeritas_objects.dir/base/ColorUtils.cc.o src/CMakeFiles/celeritas_objects.dir/base/DeviceAllocation.cc.o src/CMakeFiles/celeritas_objects.dir/comm/KernelDiagnostics.cc.o src/CMakeFiles/celeritas_objects.dir/base/ScopedStreamRedirect.cc.o src/CMakeFiles/celeritas_objects.dir/base/TypeDemangler.cc.o src/CMakeFiles/celeritas_objects.dir/base/detail/Copier.cc.o src/CMakeFiles/celeritas_objects.dir/comm/Communicator.cc.o src/CMakeFiles/celeritas_objects.dir/comm/Device.cc.o src/CMakeFiles/celeritas_objects.dir/comm/Logger.cc.o src/CMakeFiles/celeritas_objects.dir/comm/LoggerTypes.cc.o src/CMakeFiles/celeritas_objects.dir/comm/ScopedMpiInit.cc.o src/CMakeFiles/celeritas_objects.dir/comm/detail/LoggerMessage.cc.o src/CMakeFiles/celeritas_objects.dir/geometry/detail/ScopedTimeAndRedirect.cc.o src/CMakeFiles/celeritas_objects.dir/io/ImportProcess.cc.o src/CMakeFiles/celeritas_objects.dir/io/ImportPhysicsTable.cc.o src/CMakeFiles/celeritas_objects.dir/io/ImportPhysicsVector.cc.o src/CMakeFiles/celeritas_objects.dir/io/AtomicRelaxationReader.cc.o src/CMakeFiles/celeritas_objects.dir/io/LivermorePEReader.cc.o src/CMakeFiles/celeritas_objects.dir/io/SeltzerBergerReader.cc.o src/CMakeFiles/celeritas_objects.dir/physics/base/CutoffParams.cc.o src/CMakeFiles/celeritas_objects.dir/physics/base/ImportedProcessAdapter.cc.o src/CMakeFiles/celeritas_objects.dir/physics/base/Model.cc.o src/CMakeFiles/celeritas_objects.dir/physics/base/ParticleParams.cc.o src/CMakeFiles/celeritas_objects.dir/physics/base/PhysicsParams.cc.o src/CMakeFiles/celeritas_objects.dir/physics/base/Process.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/AtomicRelaxationParams.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/BetheHeitlerModel.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/ComptonProcess.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/PhotoelectricProcess.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/LivermorePEModel.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/EIonizationProcess.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/EPlusAnnihilationProcess.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/EPlusGGModel.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/GammaConversionProcess.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/KleinNishinaModel.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/MollerBhabhaModel.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/RayleighModel.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/RayleighProcess.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/detail/Rayleigh.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/SeltzerBergerModel.cc.o src/CMakeFiles/celeritas_objects.dir/physics/em/detail/Utils.cc.o src/CMakeFiles/celeritas_objects.dir/physics/grid/ValueGridBuilder.cc.o src/CMakeFiles/celeritas_objects.dir/physics/grid/ValueGridInserter.cc.o src/CMakeFiles/celeritas_objects.dir/physics/grid/ValueGridInterface.cc.o src/CMakeFiles/celeritas_objects.dir/physics/material/MaterialParams.cc.o src/CMakeFiles/celeritas_objects.dir/physics/material/detail/Utils.cc.o src/CMakeFiles/celeritas_objects.dir/random/RngInterface.cc.o src/CMakeFiles/celeritas_objects.dir/base/KernelParamCalculator.cuda.cc.o src/CMakeFiles/celeritas_objects.dir/base/detail/Filler.cu.o src/CMakeFiles/celeritas_objects.dir/physics/em/detail/BetheHeitler.cu.o src/CMakeFiles/celeritas_objects.dir/physics/em/detail/EPlusGG.cu.o src/CMakeFiles/celeritas_objects.dir/physics/em/detail/KleinNishina.cu.o src/CMakeFiles/celeritas_objects.dir/physics/em/detail/MollerBhabha.cu.o src/CMakeFiles/celeritas_objects.dir/physics/em/detail/LivermorePE.cu.o src/CMakeFiles/celeritas_objects.dir/physics/em/detail/Rayleigh.cu.o src/CMakeFiles/celeritas_objects.dir/random/detail/RngStateInit.cu.o src/CMakeFiles/celeritas_objects.dir/sim/detail/SimStateInit.cu.o src/CMakeFiles/celeritas_objects.dir/io/EventReader.cc.o src/CMakeFiles/celeritas_objects.dir/comm/DeviceIO.json.cc.o src/CMakeFiles/celeritas_objects.dir/comm/KernelDiagnosticsIO.json.cc.o src/CMakeFiles/celeritas_objects.dir/geometry/GeoMaterialParams.cc.o src/CMakeFiles/celeritas_objects.dir/geometry/GeoParams.cc.o src/CMakeFiles/celeritas_objects.dir/geometry/detail/VGNavCollection.cc.o src/CMakeFiles/celeritas_objects.dir/sim/TrackInitInterface.cc.o src/CMakeFiles/celeritas_objects.dir/sim/TrackInitParams.cc.o src/CMakeFiles/celeritas_objects.dir/sim/TrackInitUtils.cc.o src/CMakeFiles/celeritas_objects.dir/sim/detail/InitializeTracks.cu.o && /usr/bin/ranlib src/libceleritas_mid.a && :
[2/5] : && /projects/spack/opt/spack/gcc-8.3.1/cmake/547luje/bin/cmake -E rm -f app/libceleritas_demo_rasterizer_mid.a && /usr/bin/ar qc app/libceleritas_demo_rasterizer_mid.a  app/CMakeFiles/celeritas_demo_rasterizer_objects.dir/demo-rasterizer/RDemoRunner.cc.o app/CMakeFiles/celeritas_demo_rasterizer_objects.dir/demo-rasterizer/RDemoKernel.cu.o app/CMakeFiles/celeritas_demo_rasterizer_objects.dir/demo-rasterizer/ImageIO.cc.o app/CMakeFiles/celeritas_demo_rasterizer_objects.dir/demo-rasterizer/ImageStore.cc.o && /usr/bin/ranlib app/libceleritas_demo_rasterizer_mid.a && :
[3/5] /usr/local/cuda-10.2/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -g -G --generate-code=arch=compute_35,code=[compute_35,sm_35] /home/s3j/.local/src/celeritas/build-static/app/libceleritas_demo_rasterizer_mid.a /home/s3j/.local/src/celeritas/build-static/src/libceleritas_mid.a /projects/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvecgeomcuda_static.a -Xcompiler=-pthread -Xcompiler=-fPIC -Wno-deprecated-gpu-targets -shared -dlink app/CMakeFiles/celeritas_demo_rasterizer_final.dir/CMakeFiles/celeritas_demo_rasterizer_emptyfile.cu.o -o CMakeFiles/celeritas_demo_rasterizer_final.dir/cmake_device_link.o  app/libceleritas_demo_rasterizer_mid.a src/libceleritas_mid.a  -ldl -lrt -lpthread -lcudadevrt -lcudart_static
FAILED: CMakeFiles/celeritas_demo_rasterizer_final.dir/cmake_device_link.o
/usr/local/cuda-10.2/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -g -G --generate-code=arch=compute_35,code=[compute_35,sm_35] /home/s3j/.local/src/celeritas/build-static/app/libceleritas_demo_rasterizer_mid.a /home/s3j/.local/src/celeritas/build-static/src/libceleritas_mid.a /projects/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvecgeomcuda_static.a -Xcompiler=-pthread -Xcompiler=-fPIC -Wno-deprecated-gpu-targets -shared -dlink app/CMakeFiles/celeritas_demo_rasterizer_final.dir/CMakeFiles/celeritas_demo_rasterizer_emptyfile.cu.o -o CMakeFiles/celeritas_demo_rasterizer_final.dir/cmake_device_link.o  app/libceleritas_demo_rasterizer_mid.a src/libceleritas_mid.a  -ldl -lrt -lpthread -lcudadevrt -lcudart_static
nvlink error   : Multiple definition of '_ZN15demo_rasterizer70_GLOBAL__N__46_tmpxft_0011668f_00000000_6_RDemoKernel_cpp1_ii_8a0b3a9c12trace_kernelEN9celeritas13GeoParamsDataILNS1_9OwnershipE2ELNS1_8MemSpaceE1EEENS1_12GeoStateDataILS3_1ELS4_1EEENS_13ImagePointersE' in 'app/libceleritas_demo_rasterizer_mid.a:RDemoKernel.cu.o', first defined in '/home/s3j/.local/src/celeritas/build-static/app/libceleritas_demo_rasterizer_mid.a:RDemoKernel.cu.o'
nvlink error   : Multiple definition of '_ZN15demo_rasterizer70_GLOBAL__N__46_tmpxft_0011668f_00000000_6_RDemoKernel_cpp1_ii_8a0b3a9c6geo_idERKN9celeritas12GeoTrackViewE' in 'app/libceleritas_demo_rasterizer_mid.a:RDemoKernel.cu.o', first defined in '/home/s3j/.local/src/celeritas/build-static/app/libceleritas_demo_rasterizer_mid.a:RDemoKernel.cu.o'
nvlink fatal   : merge_elf failed
ninja: build stopped: subcommand failed.
```

Since we can't get this to work on emmet, but at least the CI passes and can warn us about future missing `inline` keywords, can you give this a check so we can merge?

**Comment by sethrj** [comment]
Arg keep hitting out-of-disk-space errors on the CI. Each one of these builds being 4GB really eats through the drive quickly 😠

**Comment by pcanal** [comment]
On emmet, on the failing build what does:
```
/usr/local/cuda-10.2/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -g -G --generate-code=arch=compute_35,code=[compute_35,sm_35] src/libceleritas_mid.a /projects/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvecgeomcuda_static.a -Xcompiler=-pthread -Xcompiler=-fPIC -Wno-deprecated-gpu-targets -shared -dlink app/CMakeFiles/celeritas_demo_rasterizer_final.dir/CMakeFiles/celeritas_demo_rasterizer_emptyfile.cu.o -o CMakeFiles/celeritas_demo_rasterizer_final.dir/cmake_device_link.o  app/libceleritas_demo_rasterizer_mid.a src/libceleritas_mid.a  -ldl -lrt -lpthread -lcudadevrt -lcudart_static
```
returns/prints?

**Comment by pcanal** [comment]
> Each one of these builds being 4GB really eats through the drive quickly 

I am planning to review if some of the nvcc -dlink invocation are 'spurrious'/'avoidable'.

**Comment by sethrj** [comment]
> On emmet, on the failing build what does <snip> return/print?

This may be better?
```console
$ /usr/local/cuda-10.2/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -g -G --generate-code=arch=compute_35,code=[compute_35,sm_35] src/libceleritas_mid.a /projects/spack/var/spack/environments/celeritas/.spack-env/view/lib64/libvecgeomcuda_static.a -Xcompiler=-pthread -Xcompiler=-fPIC -Wno-deprecated-gpu-targets -shared -dlink app/CMakeFiles/celeritas_demo_rasterizer_final.dir/CMakeFiles/celeritas_demo_rasterizer_emptyfile.cu.o -o CMakeFiles/celeritas_demo_rasterizer_final.dir/cmake_device_link.o  app/libceleritas_demo_rasterizer_mid.a src/libceleritas_mid.a  -ldl -lrt -lpthread -lcudadevrt -lcudart_static
nvlink error   : Multiple definition of '_ZN9celeritas6detail71_GLOBAL__N__47_tmpxft_001169ec_00000000_6_BetheHeitler_cpp1_ii_e3f3b31229bethe_heitler_interact_kernelENS0_20BetheHeitlerPointersENS_17ModelInteractRefsILNS_8MemSpaceE1EEE' in 'src/libceleritas_mid.a:BetheHeitler.cu.o', first defined in 'src/libceleritas_mid.a:BetheHeitler.cu.o'
nvlink fatal   : merge_elf failed
```

**Comment by pcanal** [comment]
Ok.  It is now clear.  The problem is indeed that the static library, somehow, appears as both a fullpathname and a relative pathname.  This disables the pass I have to remove duplicates (simple a `list(REMOVE_DUPLICATE...)`) and we end with the same static library given multiple time to `nvcc -dlink`.  To track this down, I would run `cmake --debug --trace --trace-expand .` and search the result/log for where the full path name library comes from.

Most likely, this is cmake version dependent (but could also be dependent on the PATH being used/passed).

**Comment by sethrj** [comment]
@pcanal OK now this is passing and properly fixed up! 😄

**Review by pcanal — APPROVED** [review]
Good to go, I guess.  Thanks! :)
